src/main/java/games/strategy/engine/framework/startup/login/ClientLoginValidator.java

```diff
@@ -40,6 +42,16 @@
   private final IServerMessenger serverMessenger;
   private String password;
 
+  /**
+   * Returns true if the parameter value is the correct length of a md5 hashed value,
+   * and if also it starts with the correct string value.
+   */
+  public static boolean isValidMac(final String value) {
+    return value.length() == MAC_ADDRESS_LENGTH
+            && value.startsWith(MD5Crypt.MAGIC + "MH$")
+            && value.matches("[0-9a-zA-Z$./]+");
```

Comment:
What about replacing all of those checks with a simple `value.matches("$1$MH$[0-9a-zA-Z./]{22}")`?
Haven't tested this regex but it should be something along of those lines...

Reply:
not a bad suggestion, though I'm not sure it's clearly worthwhile. Keeping the regex simpler is not a bad thing IMO.

Reply:
Plus, this is just extracted code as well, I'd rather not muck with the functionality as much (even if I did add a few tests, but those tests are not quite as comprehensive as they could be).

Again though, we are likely to be ripping out MAC as it does not seem we are getting that from network packets, it would be almost trivial to hack that and defeat the MAC banning. So we should not over-invest in this.

For now, this are some updates pulled off of a much larger branch, so I include them at all to simplify that upcoming PR.

Reply:
The point is, we can drop the length and starts with check then as well...

Reply:
Sure, but converting to one line with a more complex regex is not a clear win IMO. It's pretty easy to go through the three conditions and then only need understand the last checks for alphnum. Let's not invest any extra here, everything MAC related is likely to be 100% deleted from the code since it is almost useless (unless we can really confirm we are getting MAC from network packets, but I think that is not the case)

Reply:
> unless we can really confirm we are getting MAC from network packets, but I think that is not the case

Pretty sure you're right, @DanVanAtta.  I believe the MAC is sent as part of the application-level protocol (e.g. see `LobbyLogin#login(LoginPanel)`, where the application queries the OS for the MAC and passes it along to `ClientMessenger` to be sent as part of our custom protocol [e.g. see the `ClientQuarantineConversation` constructor where the MAC address is actually sent over the wire]).

AFAIK, you can't determine the MAC of a machine on a different segment because MACs get replaced in the data-link layer as you traverse segments (typically with the MAC of the router that forwarded the packet).  This [SO answer](https://security.stackexchange.com/a/18767/136686) provides a good overview.  Hence, why it was necessary to send the MAC at the application level.

Reply:
Excellent link @ssoloff
I think the first step to still be able to reliably ban people would be to require unique email addresses, that are actually verified. 
Only when they are verified, a user is allowed to use the chat in the lobby.

The only thing we should be aware of is disposable email services like 10minutemail.com using lists like [this one](https://gist.github.com/adamloving/4401361).
We should add an explanation for users that try to login with such a temporary address and explain them why we are not allowing them though.

Reply:
> I think the first step to still be able to reliably ban people would be to require unique email addresses, that are actually verified.
Only when they are verified, a user is allowed to use the chat in the lobby.

Grant it, this is a bit of a controversial statement. I'm not sure also if we should white list email providers, there are dozens on dozens, and some folks may very well be using a self hosted email.

When we discussesd banning and lobby chatting, there was acknowledgement it would be a loss to require registration. From that perspective our best bet is to add value to long lived accounts. To one extent, I dont' think we will have a full proof ban system, it's just too easy to get around.

Reply:
Maybe we should drop the concept of banning altogether then...
Since the only reason (I know of) for banning is that some people are just annoying, we could easily enable every player to mute everyone just for themselves, and moderators to mute everyone for everyone except for themselves (unless muted excplicitly), so people could still apologize or complain if they believe it was unjustified.
This way no registration would be required and we still can "remove" someone from the community who does not deserve to be part of it.

And if that's not enough maybe kicking someone (and/or creating a ip timeout for a day or something) will be sufficient then.

Thoughts?

Reply:
> easily enable every player to mute everyone just for themselves

We have that already, right? Moderators have an 'mute for everyone' ability as well IIRC.

Banning by MAC address I think should/will just get dropped as it is pretty broken.  The idea I think is we do want a 'kick' system with some sort of cool down before rejoining. Doing so by IP is not ideal, but it is there. MAC banning to me does not add anything beyond IP banning and is arguably even more trivially hacked than changing your IP address. Reigning in the trolls is an ongoing problem. The moderators would prefer for everyone to register and be bannable, but at the same time we can't make life more difficult for everyone just because of the 0.1% of players that are bored and want to be annoying. Grant it, at the same time we need to encourage good player reputation and behavior as much as possible to avoid that.

Reply:
> We have that already, right? Moderators have an 'mute for everyone' ability as well IIRC.

The latter one yes, not sure if everyone can mute someone "privately", will investigate in the next couple of days...